godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,777 @@
+    }
+
+    /// # Safety
+    ///
+    /// - We expect the engine to provide a valid const string name pointer.
+    unsafe fn transfer_string_name_from_godot(
+        p_name: sys::GDExtensionConstStringNamePtr,
+    ) -> StringName {
+        // This string is not ours, so we cannot call the destructor on it.
+        ManuallyDrop::new(StringName::from_string_sys(sys::force_mut_ptr(p_name)))
+            .deref()
+            .clone()
+    }
+
+    fn transfer_bool_to_godot(value: bool) -> u8 {
+        value as u8
+    }
+
+    /// # Safety
+    ///
+    /// - We expect the engine to provide a valid variant pointer the return value can be moved into.
+    unsafe fn transfer_variant_to_godot(variant: Variant, return_ptr: sys::GDExtensionVariantPtr) {
+        variant.move_var_ptr(return_ptr)
+    }
+
+    /// # Safety
+    ///
+    /// - The returned `*const T` is guaranteed to point to a list that has an equal length and capacity.
+    fn transfer_ptr_list_to_godot<T>(ptr_list: Vec<T>, list_length: *mut u32) -> *const T {
+        // SAFETY: list_length has to be a valid pointer to a u32.
+        unsafe {
+            *list_length = ptr_list.len() as u32;
+        }
+
+        let ptr = Box::into_raw(ptr_list.into_boxed_slice());
+
+        // SAFETY: `ptr` was just created in the line above and should be safe to dereference.
+        unsafe { (*ptr).as_ptr() }
+    }
+
+    fn transfer_script_to_godot(script: &Gd<crate::engine::Script>) -> sys::GDExtensionObjectPtr {
+        script.obj_sys()
+    }
+
+    /// # Safety
+    ///
+    /// - `ptr` is expected to point to a list with both a length and capacity of `list_length`.
+    /// - The list pointer `ptr` must have been created with [`transfer_ptr_list_to_godot`].
+    unsafe fn transfer_ptr_list_from_godot<T>(ptr: *const T, list_length: usize) -> Vec<T> {
+        Vec::from_raw_parts(sys::force_mut_ptr(ptr), list_length, list_length)
+    }
+
+    /// # Safety
+    ///
+    /// - The engine has to provide a valid string return pointer.
+    unsafe fn transfer_string_to_godot(string: GString, return_ptr: sys::GDExtensionStringPtr) {
+        string.move_string_ptr(return_ptr);
+    }
+
+    fn transfer_property_state_to_godot(
+        propery_states: Vec<(StringName, Variant)>,
+        property_state_add: sys::GDExtensionScriptInstancePropertyStateAdd,
+        userdata: *mut c_void,
+    ) {
+        let Some(property_state_add) = property_state_add else {
+            return;
+        };
+
+        propery_states.into_iter().for_each(|(name, value)| {
+            let name = ManuallyDrop::new(name);
+            let value = ManuallyDrop::new(value);
+
+            // SAFETY: Godot expects a string name and a variant pointer for each property. After receiving the pointer, the engine is responsible
+            // for managing the memory behind those references.
+            unsafe { property_state_add(name.string_sys(), value.var_sys(), userdata) };
```

Comment:
As this function is a safe function, you cannot assume that `property_state_add` can be called with these values. `transfer_property_state_to_godot` should be unsafe and you should mention that it must be safe to call the function in this way.

Reply:
This is because `property_state_add` will dereference the pointer to `userdata` which may not be valid?

Reply:
yes but also that function is an unsafe function so it cannot be assumed to be safe to be called in general

Reply:
ok, the function is now unsafe and I added safety documentation.